File: examples/webhook-verification/node/verify.mjs
#!/usr/bin/env node
// SPDX-License-Identifier: MIT
/**
 * OphirPay webhook signature verification — reference implementation (Node.js)
 *
 * Canonicalization (must match `buildSignedPayload` in
 * `src/lib/webhook-deliver.ts`):
 *
 *   1. Parse the received JSON body.
 *   2. Set the `signature` field to "" — keep the key, empty the value.
 *      (Do NOT delete the key; the canonical string contains `"signature":""`.)
 *   3. Re-serialize with stable key order (JSON.stringify preserves the
 *      insertion order of the parsed body, which matches the sender's order).
 *   4. Compute HMAC-SHA256 (hex) over that canonical string using your
 *      webhook secret.
 *   5. Compare against the `X-OphirPay-Signature` header with a
 *      constant-time comparison.
 *
 * CLI:
 *
 *   node verify.mjs --secret <secret> --signature <hex> \
 *     [--body-file <path>] [--max-age <seconds>] [--now <iso>]
 *
 * Reads the body from `--body-file`, or stdin when omitted. Prints "VALID"
 * and exits 0 on success, or "INVALID: <reason>" and exits 1 otherwise.
 */
import { createHmac, timingSafeEqual } from "node:crypto";
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

const DEFAULT_MAX_AGE_SECONDS = 300; // replay-protection window

/**
 * Build the canonical string a receiver must sign, byte-for-byte identical
 * to what `buildSignedPayload` signs on the sender side.
 */
export function canonicalize(body) {
  const parsed = JSON.parse(body);
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new Error("body must be a JSON object");
  }
  // Empty the signature field instead of deleting it: the HMAC input
  // includes `"signature":""` as the last key.
  return JSON.stringify({ ...parsed, signature: "" });
}

/**
 * Verify an OphirPay webhook delivery.
 *
 * @param {object} opts
 * @param {string} opts.body      raw request body (string)
 * @param {string} opts.signature value of the X-OphirPay-Signature header
 * @param {string} opts.secret    your webhook signing secret
 * @param {number} [opts.maxAgeSeconds=300] replay window; 0 disables the check
 * @param {Date}   [opts.now]     reference time (defaults to the current time)
 * @returns {{ valid: boolean, reason: string }}
 */
export function verifyWebhookSignature({
  body,
  signature,
  secret,
  maxAgeSeconds = DEFAULT_MAX_AGE_SECONDS,
  now = new Date(),
}) {
  let canonical;
  try {
    canonical = canonicalize(body);
  } catch (err) {
    return { valid: false, reason: `invalid body: ${err.message}` };
  }

  const expected = createHmac("sha256", secret).update(canonical).digest("hex");
  const provided = String(signature ?? "");
  const providedBuf = Buffer.from(provided);
  const expectedBuf = Buffer.from(expected);
  const matches =
    providedBuf.length === expectedBuf.length && timingSafeEqual(providedBuf, expectedBuf);
  if (!matches) {
    return { valid: false, reason: "signature mismatch" };
  }

  if (maxAgeSeconds > 0) {
    const parsed = JSON.parse(body);
    const ts = Date.parse(parsed.timestamp);
    if (Number.isNaN(ts)) {
      return { valid: false, reason: "missing or invalid timestamp" };
    }
    const ageSeconds = (now.getTime() - ts) / 1000;
    if (ageSeconds > maxAgeSeconds) {
      return {
        valid: false,
        reason: `payload too old (${Math.round(ageSeconds)}s > ${maxAgeSeconds}s) — possible replay`,
      };
    }
    if (ageSeconds < -maxAgeSeconds) {
      return {
        valid: false,
        reason: `payload timestamp is in the future (${Math.round(-ageSeconds)}s ahead)`,
      };
    }
  }

  return { valid: true, reason: "valid" };
}

function parseArgs(argv) {
  const args = { secret: null, signature: null, bodyFile: null, maxAge: null, now: null };
  for (let i = 0; i < argv.length; i++) {
    switch (argv[i]) {
      case "--secret":
        args.secret = argv[++i];
        break;
      case "--signature":
        args.signature = argv[++i];
        break;
      case "--body-file":
        args.bodyFile = argv[++i];
        break;
      case "--max-age":
        args.maxAge = Number(argv[++i]);
        break;
      case "--now":
        args.now = argv[++i];
        break;
      default:
        throw new Error(`unknown argument: ${argv[i]}`);
    }
  }
  return args;
}

function main() {
  let args;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (err) {
    console.error(`error: ${err.message}`);
    console.error(
      "usage: node verify.mjs --secret <secret> --signature <hex> [--body-file <path>] [--max-age <seconds>] [--now <iso>]"
    );
    process.exit(2);
  }
  if (!args.secret || !args.signature) {
    console.error(
      "usage: node verify.mjs --secret <secret> --signature <hex> [--body-file <path>] [--max-age <seconds>] [--now <iso>]"
    );
    process.exit(2);
  }
  const body = args.bodyFile ? readFileSync(args.bodyFile, "utf8") : readFileSync(0, "utf8");
  const now = args.now ? new Date(args.now) : new Date();
  const result = verifyWebhookSignature({
    body,
    signature: args.signature,
    secret: args.secret,
    maxAgeSeconds: args.maxAge ?? DEFAULT_MAX_AGE_SECONDS,
    now,
  });
  if (result.valid) {
    console.log("VALID");
    process.exit(0);
  }
  console.error(`INVALID: ${result.reason}`);
  process.exit(1);
}

// Run as a CLI when executed directly (not when imported).
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main();
}

File: examples/webhook-verification/python/verify.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
"""
OphirPay webhook signature verification - reference implementation (Python)

Signed material (must match `buildSignedPayload` in
`src/lib/webhook-deliver.ts`):

  `<timestamp>.<canonicalBody>`

  1. Take the timestamp: the `X-OphirPay-Timestamp` header value, falling
     back to the (also-signed) `timestamp` field of the body when absent.
  2. Parse the received JSON body.
  3. Set the `signature` field to "" - keep the key, empty the value.
     (Do NOT delete the key; the canonical string contains `"signature":""`.)
  4. Re-serialize with stable key order. Python 3.7+ dicts preserve
     insertion order, so parsing the body and re-serializing keeps the
     sender's key order. Use compact separators (",", ":") and raw UTF-8
     (no \\uXXXX escapes) so the bytes match Node's `JSON.stringify`.
  5. Compute HMAC-SHA256 (hex) over `<timestamp>.<canonicalBody>` using your
     webhook secret.
  6. Compare against the `X-OphirPay-Signature` header with a
     constant-time comparison (`hmac.compare_digest`).
  7. Reject a timestamp outside the freshness window (replay protection).

CLI:

  python3 verify.py --secret <secret> --signature <hex> \\
      [--timestamp <iso>] [--body-file <path>] [--max-age <seconds>] [--now <iso>]

Reads the body from `--body-file`, or stdin when omitted. Prints "VALID"
and exits 0 on success, or "INVALID: <reason>" and exits 1 otherwise.
"""

import argparse
import hashlib
import hmac
import json
import sys
from datetime import datetime, timezone

DEFAULT_MAX_AGE_SECONDS = 300  # replay-protection window


def canonicalize(body):
    """Build the canonical string a receiver must sign, byte-for-byte
    identical to what `buildSignedPayload` signs on the sender side."""
    parsed = json.loads(body)
    if not isinstance(parsed, dict):
        raise ValueError("body must be a JSON object")
    # Empty the signature field instead of deleting it: the HMAC input
    # includes `"signature":""` as the last key.
    parsed["signature"] = ""
    # separators=(",", ":") and ensure_ascii=False match Node's
    # `JSON.stringify` byte-for-byte for JSON-safe payloads.
    return json.dumps(parsed, separators=(",", ":"), ensure_ascii=False)


def verify_webhook_signature(
    body,
    signature,
    secret,
    timestamp=None,
    max_age_seconds=DEFAULT_MAX_AGE_SECONDS,
    now=None,
):
    """Verify an OphirPay webhook delivery.

    Returns (valid: bool, reason: str). `timestamp` is the value of the
    X-OphirPay-Timestamp header; when omitted the body's timestamp field
    (which is signed too) is used.
    """
    try:
        parsed = json.loads(body)
        canonical = canonicalize(body)
    except (ValueError, json.JSONDecodeError) as exc:
        return False, "invalid body: %s" % exc

    # The timestamp is authenticated: it is prepended to the canonical body.
    signed_timestamp = str(
        timestamp if timestamp is not None else parsed.get("timestamp", "")
    )
    signed_input = "%s.%s" % (signed_timestamp, canonical)

    expected = hmac.new(
        secret.encode("utf-8"), signed_input.encode("utf-8"), hashlib.sha256
    ).hexdigest()
    provided = (signature or "").encode("utf-8")
    expected_bytes = expected.encode("utf-8")
    if len(provided) != len(expected_bytes) or not hmac.compare_digest(
        provided, expected_bytes
    ):
        return False, "signature mismatch"

    if max_age_seconds > 0:
        try:
            ts = datetime.fromisoformat(signed_timestamp.replace("Z", "+00:00"))
        except (ValueError, TypeError):
            return False, "missing or invalid timestamp"
        if ts.tzinfo is None:
            ts = ts.replace(tzinfo=timezone.utc)
        now = now or datetime.now(timezone.utc)
        age_seconds = (now - ts).total_seconds()
        if age_seconds > max_age_seconds:
            return False, "payload too old (%ds > %ds) - possible replay" % (
                round(age_seconds),
                max_age_seconds,
            )
        if age_seconds < -max_age_seconds:
            return False, "payload timestamp is in the future (%ds ahead)" % round(
                -age_seconds
            )

    return True, "valid"


def main(argv=None):
    parser = argparse.ArgumentParser(description="Verify an OphirPay webhook signature.")
    parser.add_argument("--secret", required=True, help="your webhook signing secret")
    parser.add_argument(
        "--signature", required=True, help="value of the X-OphirPay-Signature header"
    )
    parser.add_argument(
        "--timestamp",
        help="value of the X-OphirPay-Timestamp header (defaults to the body timestamp)",
    )
    parser.add_argument(
        "--body-file", help="path to the received JSON body (defaults to stdin)"
    )
    parser.add_argument(
        "--max-age",
        type=int,
        default=DEFAULT_MAX_AGE_SECONDS,
        help="replay window in seconds (0 disables)",
    )
    parser.add_argument(
        "--now", help="reference timestamp (ISO 8601); defaults to the current time"
    )
    args = parser.parse_args(argv)

    if args.body_file:
        with open(args.body_file, "r", encoding="utf-8") as fh:
            body = fh.read()
    else:
        body = sys.stdin.read()

    now = None
    if args.now:
        now = datetime.fromisoformat(args.now.replace("Z", "+00:00"))
        if now.tzinfo is None:
            now = now.replace(tzinfo=timezone.utc)

    valid, reason = verify_webhook_signature(
        body,
        args.signature,
        args.secret,
        timestamp=args.timestamp,
        max_age_seconds=args.max_age,
        now=now,
    )
    if valid:
        print("VALID")
        return 0
    print("INVALID: %s" % reason, file=sys.stderr)
    return 1


if __name__ == "__main__":
    sys.exit(main())
